src/overlays/actors/ovl_En_Kusa/z_en_kusa.c

```diff
@@ -55,72 +110,645 @@ static InitChainEntry D_80936754[] = {
+    s32 pad;
+    s32 kusaType = GET_KUSA_TYPE(this);
+
+    Actor_ProcessInitChain(&this->actor, sInitChain);
+
+    if (globalCtx->csCtx.state != 0) {
+        this->actor.uncullZoneForward += 1000.0f;
+    }
+    EnKusa_InitCollider(&this->actor, globalCtx);
+    CollisionCheck_SetInfo(&this->actor.colChkInfo, NULL, &sColChkInfoInit);
+
+    if (kusaType == ENKUSA_TYPE_BUSH) {
+        this->actor.shape.shadowScale = 1.0f;
+        this->actor.shape.shadowAlpha = 0x3C;
+    } else {
+        this->actor.shape.shadowScale = 0.9f;
+        this->actor.shape.shadowAlpha = 0x46;
+    }
+
+    if (this->actor.shape.rot.y == 0) {
+        this->actor.shape.rot.y = ((u32)Rand_Next() >> 0x10);
+        this->actor.home.rot.y = this->actor.shape.rot.y;
+        this->actor.world.rot.y = this->actor.shape.rot.y;
+    }
+    if (EnKusa_SnapToFloor(this, globalCtx, 0.0f) == 0) {
+        Actor_MarkForDeath(&this->actor);
+        return;
+    }
+    if (EnKusa_GetWaterBox(this, globalCtx) != 0) {
+        this->unk198 |= 1;
+    }
+
+    this->objIndex = Object_GetIndex(&globalCtx->objectCtx, objectIds[(this->actor.params & 3)]);
+    if (this->objIndex < 0) {
+        Actor_MarkForDeath(&this->actor);
+        return;
+    }
+
+    EnKusa_SetupWaitObject(this);
+    if (D_809366B4) {
+        D_80936CD8 = ((u32)Rand_Next() >> 0x10);
+        D_80936CDA = ((u32)Rand_Next() >> 0x10);
+        D_80936CDC = ((u32)Rand_Next() >> 0x10);
+        D_80936CDE = ((u32)Rand_Next() >> 0x10);
+        D_80936CE0 = ((u32)Rand_Next() >> 0x10);
+        D_809366B4 = false;
+        EnKusa_Sway();
+        kusaGameplayFrames = globalCtx->gameplayFrames;
+    }
+    this->kusaMtxIdx = D_809366B0 & 7;
+    D_809366B0 += 1;
+}
+
+void EnKusa_Destroy(Actor* thisx, GlobalContext* globalCtx) {
+    GlobalContext* globalCtx2 = globalCtx;
+    EnKusa* this = THIS;
+
+    Collider_DestroyCylinder(globalCtx, &this->collider);
+}
+
+void EnKusa_SetupWaitObject(EnKusa* this) {
+    this->actionFunc = EnKusa_WaitObject;
+}
+
+/**
+ * @brief Waits for object to be loaded then draws the appropriate actor
+ *
+ */
+void EnKusa_WaitObject(EnKusa* this, GlobalContext* globalCtx) {
+    s16 pad;
+
+    if (Object_IsLoaded(&globalCtx->objectCtx, this->objIndex) != 0) {
+        s32 kusa_type = GET_KUSA_TYPE(this);
+        if (this->isCut != 0) {
+            EnKusa_SetupCut(this);
+        } else {
+            EnKusa_SetupMain(this);
+        }
+        if (kusa_type == ENKUSA_TYPE_BUSH) {
+            this->actor.draw = EnKusa_DrawBush;
+        } else {
+            this->actor.draw = EnKusa_DrawGrass;
+        }
+        this->actor.objBankIndex = this->objIndex;
+        this->actor.flags &= ~0x10;
+    }
+}
+
+void EnKusa_SetupMain(EnKusa* this) {
+    this->actionFunc = EnKusa_Main;
+    this->actor.flags &= ~0x10;
+}
+
+/**
+ * @brief Handles the main logic and control flow for all kusa types
+ *
+ */
+void EnKusa_Main(EnKusa* this, GlobalContext* globalCtx) {
```

Comment:
I don't really like Main as a name for an actor function, it's a bit too GameState-y.